ui: make rows in jobs table expandable to see details

vilterp opened the pull request:
Fixes #20990

![image](https://user-images.githubusercontent.com/7341/46769475-eeaaff00-ccb9-11e8-9d7b-3cabe559b312.png)

Previously, there was no way to see job errors, even though they were loaded into JS memory. This also gets rid of the `ExpandableString`, which caused layout jitter that was somewhat awkward.

cc @j-low

Commit:
ui: allow SortedTable and SortableTable to expand and collapse rows

if optional "expandable config" is provided.

Release note: None

Commit:
ui: jobs page: use expandable table

Release note (admin ui change): Update jobs table to make each row
expandable, allowing the user to see the error message for failed jobs.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/29450)
<!-- Reviewable:end -->

j-low commented:
@vilterp this looks great.  I think this is almost certainly the route to go-- much better UX than trying to shoehorn the failures into a table cell or a tooltip.

danhhz commented:
HALLELUJAH

mjibson commented on pkg/ui/src/views/jobs/index.tsx:
```diff
@@ -226,6 +225,23 @@ class JobsTable extends React.Component<JobsTableProps, {}> {
+        <h3>Command</h3>
+        <pre className="job-detail">{job.description}</pre>
+
+        {/*TODO(vilterp): only show if there is an error*/}
```
this todo looks like it has been done

mjibson commented on pkg/ui/src/views/shared/components/sortabletable/index.tsx:
```diff
@@ -96,11 +118,28 @@ export class SortableTable extends React.Component<TableProps, {}> {
+        className="sort-table__cell sort-table__cell__expansion-control"
+        onClick={() => expandableConfig.onChangeExpansion(rowIndex, !expanded)}
+      >
+        <div style={{ width: 15 }}>
```
the width dosen't need a unit?

madelynnblue reviewed (commented):
this looks pretty cool. am excited to have it.

mrtracy reviewed (approved):
<img class="emoji" title=":LGTM:" alt=":LGTM:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> with suggestion, good implementation.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/29450#-:-LL_qICh98AAA_SFpHcf:bi5msbg)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale)

___
*[pkg/ui/src/views/shared/components/sortedtable/index.tsx, line 155 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/29450#-LL_pxqV-bSBgYpQ3GfS:-LL_pxqV-bSBgYpQ3GfT:b-6y3nx0) ([raw file](https://github.com/cockroachdb/cockroach/blob/752ba982292330ec6e2739e45b909927d7e8d6b9/pkg/ui/src/views/shared/components/sortedtable/index.tsx#L155)):*
> ```XML
>     const key = this.getKeyAt(rowIndex);
>     const expandedRows = this.state.expandedRows;
>     // TODO(vilterp): should I be using an immutable array or something
> ```

Just do `const expandedRows = this.state.expandeRows.slice(0)` and call it a day!


<!-- Sent from Reviewable.io -->

vilterp commented:
Designs from @josueeee https://www.figma.com/file/DutfNPhqFRZJw6vFcAMUjVFW/Jobs?node-id=0%3A1

Commit:
ui: lint: fix inconsistent indentation in sqlhighlight.styl

Release note: None

Commit:
ui: jobs page: remove background colors based on job state

With the icons, these add more visual distraction than they're worth.

Release note: None

Commit:
ui: jobs page: add icons to indicate job state

Green checkmark for completed, red exclamation point for failed.

Release note: None

Commit:
ui: lint: fix indentation in jobs page

To use

```
return (
  <div>
    ...
  </div>
);
```

Instead of

```
return <div>
 ...
</div>;
```

Release note: None

vilterp commented:
@couchand this should be good to go now. PTAL.

couchand commented on pkg/ui/src/views/jobs/index.tsx:
```diff
@@ -148,9 +175,8 @@ const jobsTableColumns: ColumnDescriptor<Job>[] = [
   {
     title: "Description",
-    cell: job => <ExpandableString long={job.description} />,
+    cell: job => (<div className="jobs-table__cell--description">{job.description}</div>),
```
nit: no parens for one-liners.

vilterp replied:
Done

couchand commented on pkg/ui/src/views/shared/components/sortedtable/index.tsx:
```diff
@@ -120,16 +141,71 @@ export class SortedTable<T> extends React.Component<SortedTableProps<T>, {}> {
+    super(props);
+    // TODO(vilterp): use a LocalSetting instead so the expansion state
+    // will persist if the user navigates to a different page and back.
+    this.state = {
```
We can get rid of the constructor override by writing the init this way: https://github.com/cockroachdb/cockroach/blob/304150d8010b05cd5bd106e6e6b63a89c2835403/pkg/ui/src/views/shared/components/expandableString/index.tsx#L26

couchand commented on pkg/ui/src/views/shared/components/sortedtable/index.tsx:
```diff
@@ -45,6 +50,22 @@ interface SortedTableProps<T> {
+}
+
+interface SortedTableState {
+  expandedRows: Set<string>;
```
It looks like the changes at this level are mostly about keeping state for the sortable table below.  The pattern to separate state-keeping containers from stateless components makes sense, but I'm not sure if the state belongs here.  In this case it splits the logic and data structures (there are two `expandableConfig` props with similar but different keys, for instance).  It seems like it would be clearer either inline in sortable table itself or in a thin wrapper in the same file.

vilterp replied:
The `expandableConfig` thing is annoying, but it still seems to make sense to keep the expansion state next to the sort state, in `SortedTable`. Not quite following your suggestion.

couchand commented on pkg/ui/src/views/shared/components/sortabletable/index.tsx:
```diff
@@ -128,16 +155,49 @@ export class SortableTable extends React.Component<TableProps, {}> {
+            <tr
+              key={rowIndex}
+              className={classes}
+              onClick={() => expandableConfig.onChangeExpansion(rowIndex, !expanded)}
```
Seems like this should be defensive to null.

vilterp replied:
Done

couchand commented on pkg/ui/src/views/shared/components/sortabletable/index.tsx:
```diff
@@ -128,16 +155,49 @@ export class SortableTable extends React.Component<TableProps, {}> {
+              "sort-table__row",
+              "sort-table__row--body",
+              "sort-table__row--expanded-area",
+            );
```
I wonder if this should mirror `classes` and make a call to `this.props.rowClass(rowIndex)`?

It looks like that prop is only used on this page, to set the job status class to match rules in `styl/pages/jobs.styl`.  Worth looking at the rules that remain.

couchand commented on pkg/ui/src/views/shared/components/sortabletable/index.tsx:
```diff
@@ -128,16 +155,49 @@ export class SortableTable extends React.Component<TableProps, {}> {
             this.props.rowClass(rowIndex),
+            { "sort-table__row--expandable": !!expandableConfig },
           );
-          return (
```
This per row function is getting a little unwieldy inline here in render.  Maybe worth pulling it out.

vilterp replied:
Done

couchand commented on pkg/ui/src/views/shared/components/sortabletable/index.tsx:
```diff
@@ -96,11 +111,23 @@ export class SortableTable extends React.Component<TableProps, {}> {
   }
 
+  expansionControl(expanded: boolean) {
+    const content = expanded ? "▼" : "▶";
```
If these were in the stylesheet they'd be right next to the up/down arrows, which seems nice.

vilterp replied:
Yeah, but it seems more straightforward just to put them in the code, rather than in `:after`… I dunno

couchand reviewed (commented):
One bigger question about sortabletable/sortedtable, and a few nits/conversation starters.

@josueeee have you had a chance to take a look?

vilterp commented:
Note: gonna fix #30682 while we're at it. cc @eriktrinh

vilterp commented on pkg/ui/src/views/jobs/index.tsx:
```diff
@@ -70,36 +73,60 @@ const formatDuration = (d: moment.Duration) =>
+    if (this.is(JOB_STATUS_SUCCEEDED, JOB_STATUS_FAILED, JOB_STATUS_CANCELLED)) {
+      return (
+        <div className="jobs-table__status">
+          {this.props.job.status !== JOB_STATUS_CANCELLED
```
Thanks; knew there was a better way to write this.

Commit:
ui: jobs page: update color of progress bar

To match palette

Release note: None

Commit:
ui: refactor: pages/jobs.styl => jobs/index.styl

Release note: None

vilterp commented:
This was LGTM'd in Slack by @josueeee 

@couchand PTAL

Commit:
ui: jobs page: capitalize running status

Release note: None

couchand reviewed (dismissed):
> Reviewed 3 of 5 files at r13, 4 of 9 files at r14, 2 of 5 files at r17, 1 of 3 files at r21, 2 of 2 files at r22.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/29450)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale)

___
*[pkg/ui/src/views/shared/components/sortabletable/index.tsx, line 165 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/29450#-LLH1uJSEK6RmdgKWvxs-r9-165:-LOVkg9K1kTKixtXx_fz:b202lyt) ([raw file](https://github.com/cockroachdb/cockroach/blob/81245f47e853025aac7d9a55d99c7e254a4aa4db/pkg/ui/src/views/shared/components/sortabletable/index.tsx#L165)):*
<details><summary><i>Previously, vilterp (Pete Vilter) wrote…</i></summary><blockquote>

Done
</blockquote></details>

Looks like it's still susceptible to it.  If `expandableConfig` is null, `onClickExpand` will be null.
___
*[pkg/ui/src/views/shared/components/sortabletable/sortabletable.styl, line 54 at r22](https://reviewable.io/reviews/cockroachdb/cockroach/29450#-LOVkSq-8WK9fYuuMfDE:-LOVkSq05kKQiaKo4K96:b9cs1m1) ([raw file](https://github.com/cockroachdb/cockroach/blob/f77a4c46a3cadc721e89ac7a65e3c48f02af7725/pkg/ui/src/views/shared/components/sortabletable/sortabletable.styl#L54)):*
> ```Stylus
> 
>     &--expanded-area
>       background-color #E2E5EE !important // TODO(vilterp): palette-ize
> ```

and de-importantify!
___
*[pkg/ui/src/views/shared/components/sortedtable/index.tsx, line 68 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/29450#-LLH1uJSEK6RmdgKWvxu-r9-68:-LOVlfAB7BESGlO32kHI:bg3dju2) ([raw file](https://github.com/cockroachdb/cockroach/blob/81245f47e853025aac7d9a55d99c7e254a4aa4db/pkg/ui/src/views/shared/components/sortedtable/index.tsx#L68)):*
<details><summary><i>Previously, vilterp (Pete Vilter) wrote…</i></summary><blockquote>

The `expandableConfig` thing is annoying, but it still seems to make sense to keep the expansion state next to the sort state, in `SortedTable`. Not quite following your suggestion.
</blockquote></details>

I suppose my beef is with `SortableTable`/`SortedTable` generally.  Now it's `SortableAndExpandableTable` vs. `SortedAndExpandedTable`.  Fine for the time being, but let's revisit this structure again soon.


<!-- Sent from Reviewable.io -->

couchand commented:
LGTM none of those comments are pressing

couchand commented:
bors r+

vilterp commented:
bors r+
